eessi/testsuite/tests/apps/osu.py

```diff
@@ -136,24 +136,16 @@ def set_num_gpus_per_node(self):
         This test does not require gpus and is for host to host within GPU nodes. But some systems do require a GPU
         allocation for to perform any activity in the GPU nodes.
         """
-        if(FEATURES[GPU] in self.current_partition.features and not utils.is_cuda_required_module(self.module_name)):
-            max_avail_gpus_per_node = utils.get_max_avail_gpus_per_node(self)
-            # Here for the 2_node test we assign max_avail_gpus_per_node but some systems cannot allocate 1_cpn_2_nodes
-            # for GPUs but need all gpus allocated within the 2 nodes for this work which. The test may fail under such
-            # conditions for the scale 1_cpn_2_nodes because it is simply not allowed.
-            self.num_gpus_per_node = self.default_num_gpus_per_node or max_avail_gpus_per_node
-        elif(FEATURES[GPU] in self.current_partition.features and utils.is_cuda_required_module(self.module_name)):
-            max_avail_gpus_per_node = utils.get_max_avail_gpus_per_node(self)
-            if(SCALES.get(self.scale).get('num_nodes') == 1):
-                # Skip the single node test if there is only 1 device in the node.
-                if(max_avail_gpus_per_node == 1):
-                    self.skip(msg="There is only 1 device within the node. Skipping tests involving only 1 node.")
-                else:
-                    self.num_gpus_per_node = 2
-            else:
-                # Note these settings are for 1_cpn_2_nodes. In that case we want to test for only 1 GPU per node since
-                # we have not requested for full nodes.
-                self.num_gpus_per_node = self.default_num_gpus_per_node or max_avail_gpus_per_node
+        if self.device_type == DEVICE_TYPES[GPU]:
+            # Skip scales with only 1 GPU device and single-node tests with only 1 GPU device in the node
+            self.skip_if(
+                SCALES[self.scale]['num_nodes'] == 1 and self.default_num_gpus_per_node == 1,
```

Comment:
`SCALES[self.scale]['num_nodes'] == 1` we already know very early on. I'd prefer to _not_ generate a test for that at all (keep the amount of generated & skipped test to a minimum - this was our whole reason for using `features` as well  :)). I.e. check for this in the `init` phase, and simply put `valid_systems=[]` or something.

For `self.default_num_gpus_per_node`, you can ony check after the setup, since that (potentially) depends on the amount of GPUs available per node for that partition. So I agree, then you cannot do anything but skip it after the setup phase. However, I expect that to be a more rare occurence (especially on HPC systems), so shouldn't give too many skipped tests...

Reply:
To clarify: 

I mean skipping the scale `'2_cores': {'num_nodes': 1, 'num_cpus_per_node': 2, 'num_gpus_per_node': 1},` when `self.device_type == DEVICE_TYPES[GPU]` can be done in the `init` phase. Your condition here won't change, but will only be left to catch single node tests when only 1 physical GPU is present in the node (but isn't that already skipped by https://github.com/EESSI/test-suite/blob/00c4bd3b897d5713b4d0a178ab305a6c405575ad/eessi/testsuite/hooks.py#L246 by the way?)

Reply:
@casparvl the test is only skipped if `test.default_num_gpus_per_node > max_avail_gpus_per_node` which is not the case always such as in Snellius.

Reply:
indeed, the hook you linked to only skips if the number of gpus per node as defined by the scale is larger than the actual number of gpus present in the node.